### DISCUSSION TITLE
Sundials upgrade in CasADi

### DISCUSSION THREAD

**Discussion by dhathris** [casadi/casadi#3954] [Q&A]
Dear CasADi developers, do you have any plans to upgrade Sundials version used in CasADi from 2.6.1 to the latest (8.1) any time soon?

**Comment by jaeandersson** [comment]
This are no such plans at the moment. There might be reasons to revisit it in 2025 if it turns out useful in any of the projects we're involved in.

**Comment by rolandmas** [comment]
Hi, Roland Mas here. I'm a Debian developer and freelance consultant, and I've been tasked by the fine folks at Synchrotron SOLEIL with preparing Debian packages for CADquery, which has a dependency on CasADi, so I'm also working on packaging CasADi. I've noticed that the casadi tarball embeds copies of external libraries, including Sundials, which is called "vendoring" and frowned upon in Debian circles, especially when such vendored libraries are already present in Debian.
Unfortunately, just trying to link against the Sundials library in Debian leads to failure, since CasADi expects its own ancient version and is no longer compatible with more current ones.

```
cd /build/casadi-3.7.0/obj-x86_64-linux-gnu/casadi/interfaces/sundials && /usr/bin/c++ -DCASADI_DEFAULT_COMPILER_PLUGIN=shell -DCASADI_IS_RELEASE=1 -DCASADI_MAJOR_VERSION=3 -DCASADI_MINOR_VERSION=7 -DCASADI_PATCH_VERSION=0 -DCASADI_SNPRINTF=snprintf -DCASADI_VERSION=31 -DHAVE_MKSTEMPS -DUSE_CXX11 -DWITH_DEEPBIND -DWITH_DEPRECATED_FEATURES -DWITH_DL -DWITH_SYSTEM_SUNDIALS -D_SCL_SECURE_NO_WARNINGS -D_USE_MATH_DEFINES -Dcasadi_sundials_common_EXPORTS -I/build/casadi-3.7.0/. -I/build/casadi-3.7.0/obj-x86_64-linux-gnu -I/build/casadi-3.7.0/casadi/interfaces/sundials/.. -I/build/casadi-3.7.0/obj-x86_64-linux-gnu/casadi/core/runtime -I/build/casadi-3.7.0/casadi/core/../.. -I/build/casadi-3.7.0/obj-x86_64-linux-gnu/casadi/core/../.. -I/usr/include/x86_64-linux-gnu/openmpi -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC  -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -MD -MT casadi/interfaces/sundials/CMakeFiles/casadi_sundials_common.dir/sundials_interface.cpp.o -MF CMakeFiles/casadi_sundials_common.dir/sundials_interface.cpp.o.d -o CMakeFiles/casadi_sundials_common.dir/sundials_interface.cpp.o -c /build/casadi-3.7.0/casadi/interfaces/sundials/sundials_interface.cpp
/build/casadi-3.7.0/casadi/interfaces/sundials/sundials_interface.cpp: In member function 'virtual int casadi::SundialsInterface::init_mem(void*) const':
/build/casadi-3.7.0/casadi/interfaces/sundials/sundials_interface.cpp:298:26: error: too few arguments to function '_generic_N_Vector* N_VNew_Serial(sunindextype, SUNContext)'
  298 |   m->v_xz = N_VNew_Serial(nx_ + nz_);
      |             ~~~~~~~~~~~~~^~~~~~~~~~~
In file included from /build/casadi-3.7.0/casadi/interfaces/sundials/sundials_interface.hpp:32,
                 from /build/casadi-3.7.0/casadi/interfaces/sundials/sundials_interface.cpp:26:
/usr/include/nvector/nvector_serial.h:90:10: note: declared here
   90 | N_Vector N_VNew_Serial(sunindextype vec_length, SUNContext sunctx);
      |          ^~~~~~~~~~~~~

```

It would really be useful if CasADi were upgraded to a recent version of Sundials (and possibly other external_packages, I haven't checked yet); I could then just link against the version present in Debian and proceed with packaging.

@jaeandersson Could you reconsider your decision? This would really help get CasADi into Debian.

Thanks!

**Reply by jaeandersson** [reply]
> @jaeandersson Could you reconsider your decision? This would really help get CasADi into Debian.

No decision has been made not to upgrade Sundials in CasADi. It's just that no one has provided time of funding to have the upgrade done. It's unclear to me how much work it would be and what the benefit would be, given that CasADi's Sundials interface uses its own structure exploiting linear solver interface and relies on CasADi's linear solvers.

**Reply by brunoml5** [reply]
Is SUNDIALS any good without the linear solver you guys implemented? Is it appropriate for the type of problems we usually find in optimal control? I'm just curious about how the tools interact, it is not important

**Reply by jaeandersson** [reply]
It's a good solver, but the forward/adjoint sensitivity analysis in CasADi is only efficient with the structure-exploiting solver we implemented. So the native solvers are not very useful IMO. CasADi integration problems are also typically small- to medium size, not huge PDEs as SUNDIALS as been designed to also work well with.

**Comment by jgillis** [comment]
This sounds like a great reason to spend some time updating. However, have you determined that CADquery needs sundials? CasADi has a massive amount of optional dependencies (we call them plugins). Chances are you could satisfy CADquery needs with only a few, or even none at all.

**Reply by rolandmas** [reply]
Good point. I'll have a look, thanks for the tip!

**Reply by jgillis** [reply]
I had a quick look at https://github.com/CadQuery/cadquery/blob/b94a06799ab36d62f9faa20021ddba3d1d4a2ecb/cadquery/occ_impl/solver.py#L17. Pretty sure you just need ipopt plugin and nothing else..

**Reply by rolandmas** [reply]
Great. So I uploaded CasADi without the external_packages/* modules for now (but I still link with Debian's libtinyxml2, libdsdp and libipopt since they are available). As far as I'm concerned the need is over, but I'll keep monitoring this ticket. If CasADi updates its sundials copy or is able to link to an up-to-date external lib, then I'll enable that option too, it might be useful to Debian users besides CADquery.
Thanks all for the help!
